src/submission/exceptions/exceptions.ts

```diff
@@ -0,0 +1,199 @@
+ * @param programId
+ * @returns program and donor level exceptions for this programId
+ */
+const queryForExceptions = async (programId: string) => {
+  const programException = await programExceptionRepository.find(programId);
+  const entityException = await entityExceptionRepository.find(programId);
+  return { programException, entityException };
+};
+
+/**
+ * Checks if there is a program exception or entity exception matching the record value
+ *
+ * @param exceptions
+ * @param validationError
+ * @param fieldValue
+ * @returns true if an exception match exists, false otherwise
+ */
+export const validateFieldValueWithExceptions = ({
+  programException,
+  entityException,
+  schemaName,
+  fieldValue,
+  validationErrorFieldName,
+}: {
+  programException: ProgramException | RepoError;
+  entityException: EntityException | RepoError;
+  schemaName: ClinicalEntitySchemaNames;
+  fieldValue: string;
+  validationErrorFieldName: string;
+}): boolean => {
+  let exceptionValue: string | undefined = '';
+  // program level is applicable to ALL donors
+  if (isProgramException(programException)) {
+    exceptionValue = programException.exceptions.find(exception => {
+      exception.requested_core_field === validationErrorFieldName;
+    })?.requested_core_field;
+  } else if (isEntityException(entityException)) {
+    const exceptionSchemaName = mapClinicalEntityNameToExceptionName(schemaName);
+    if (exceptionSchemaName) {
+      const exceptions: Array<ExceptionRecord> = entityException[exceptionSchemaName];
+      exceptionValue = exceptions.find(
+        exception => exception.requested_core_field === validationErrorFieldName,
+      )?.requested_core_field;
+    } else {
+      return false;
+    }
+  }
+  // check submitted exception value matches record validation error field value
+  return exceptionValue === fieldValue;
+};
+
+/**
+ * Normalizes input string to start with Upper case, remaining
+ * characters lowercase and to trim whitespace
+ *
+ * @param value
+ * @returns normalized string
+ */
+export const normalizeExceptionValue = (value: string) => _.upperFirst(value.trim().toLowerCase());
+
+/**
+ * map uploaded clinical type schema name with underscores to exception schema name camel cased
+ * Partial<> until all donor entities are accounted for
+ * @param schemaName
+ */
+
+const clinicalEntities: Partial<Record<
+  ClinicalEntitySchemaNames,
+  Exclude<keyof EntityException, 'programId'>
+>> = {
+  // donor: 'donor',
+  specimen: 'specimen',
+  //   primary_diagnosis: 'primaryDiagnoses',
+  //   family_history: 'familyHistory',
+  //   treatment: 'treatment',
+  //   chemotherapy: 'chemotherapy',
+  //   immunotherapy: 'immunotherapy',
+  //   surgery: 'surgery',
+  //   radiation: 'radiation',
+  //   follow_up: 'followUps',
+  //   hormone_therapy: 'hormoneTherapy',
+  //   exposure: 'exposure',
+  //   comorbidity: 'comorbidity',
+  //   biomarker: 'biomarker',
+  //   sample_registration: 'sampleRegistration',
+};
```

Comment:
a lot of overlap with `EntityAlias` in common model (https://github.com/icgc-argo/argo-clinical/blob/develop/src/common-model/entities.ts#L44) but the lack of consistency makes it a bad choice ie. the `EntityAlias` can be `primaryDiagnoses` or `primary_diagnoses` when we specifically want to map the upload clinical type to the schema name exceptions use

Reply:
They serve different purposes, at least for now. 
It would be great to revisit EntityAlias and aliasEntityNames and improve their types. 
One of the main functions for those objects is to match Donor keys w/ Entity Names, such as matching `specimen` (Clinical Entity Name) with `specimens` (keyof Donor), or 'treatment' w/ 'treatments'.

So if we wanted we could look at aligning Donors, Exceptions and Schema Names. 
Iirc those inconsistencies are coming from a handful of places and it may not be so straightforward.